File: clients/Ruby/model/build_action.rb
require_relative 'entity_type'
require_relative 'vec2_int'
class BuildAction
    attr_accessor :entity_type
    attr_accessor :position
    def initialize(entity_type, position)
        @entity_type = entity_type
        @position = position
    end
    def self.read_from(stream)
        entity_type = stream.read_int()
        if entity_type < 0 || entity_type > 10
            raise "Unexpected tag value"
        end
        position = Vec2Int.read_from(stream)
        BuildAction.new(entity_type, position)
    end
    def write_to(stream)
        stream.write_int(@entity_type)
        @position.write_to(stream)
    end
end

File: clients/TypeScript/src/model/BuildAction.ts
import {Vec2Int} from './Vec2Int';
import {EntityType} from './EntityType';
import {Stream} from './Stream';

export class BuildAction {
    constructor(private entityType: EntityType, private position: Vec2Int) {}

    static async readFrom(stream: Stream) {
        const entityType = await stream.readInt();
        const position = await Vec2Int.readFrom(stream);

        return new BuildAction(entityType, position);
    }

    async writeTo(stream: Stream) {
        await stream.writeInt(this.entityType);
        await this.position.writeTo(stream);
    }
}
